R/lang.R

```diff
@@ -206,76 +253,71 @@ is_binary_call <- function(x, name = NULL) {
 
 #' Coerce an object to a name or call.
 #'
-#' These are a S3 generics with built-in methods for names, calls, formuals,
-#' and strings. The distinction between a name and a call is particularly
-#' important when coercing from a string. Coercing to a call will parse the
-#' string, coercing to a name will create a (potentially) non-syntactic name.
+#' These coercing functions can transform names, calls, formulas, and
+#' strings. The distinction between a name and a call is particularly
+#' important when coercing from a string. Coercing to a call will
+#' parse the string, coercing to a name will create a (potentially)
+#' non-syntactic name.
 #'
 #' @param x An object to coerce
 #' @export
+#' @return \code{as_symbol()} and \code{as_call()} return a symbol or
+#'   a call. \code{as_name()} returns a string.
 #' @examples
-#' as_name("x + y")
+#' as_symbol("x + y")
 #' as_call("x + y")
 #'
 #' as_call(~ f)
-#' as_name(~ f())
-as_name <- function(x) {
-  UseMethod("as_name")
-}
-#' @export
-as_name.name <- function(x) {
-  x
-}
-#' @export
-as_name.character <- function(x) {
-  if (!is_scalar_character(x)) {
-    abort("Cannot parse character vector of length > 1")
-  }
-  symbol(x)
-}
-#' @export
-as_name.call <- function(x) {
-  if (is_prefixed_name(x)) {
-    x
-  } else {
-    as_name(x[[1]])
-  }
+#' as_symbol(~ f())
+as_symbol <- function(x) {
+  switchpatch(x, .to = "symbol",
+    symbol = x,
+    string = symbol(x),
+    quote = as_symbol(f_rhs(x)),
+    language =
+      if (is_prefixed_name(x)) {
+        x
```

Comment:
yes that isn't ideal. It was done so we could pattern-match calls with namespaced symbols:

```r
expr <- quote(base::list(x, y))
is_call(expr, quote(base::list))
```

We probably need a better way.